# DATA 512: Human-Centered Data Science
## Homework 2: Considering Bias in Data

The goal of this assignment is to explore the concept of bias in data using Wikipedia articles. This assignment will consider articles on political figures from different countries. For this assignment, you will combine a dataset of Wikipedia articles with a dataset of country populations, and use a machine learning service called ORES to estimate the quality of each article.

You are expected to perform an analysis of how the coverage of politicians on Wikipedia and the quality of articles about politicians varies among countries. Your analysis will consist of a series of tables that show:

1. The countries with the greatest and least coverage of politicians on Wikipedia compared to their population.
2. The countries with the highest and lowest proportion of high quality articles about politicians.
3. A ranking of geographic regions by articles-per-person and proportion of high quality articles.


# 0. Setup

Importing required packages, and setting up functions required for analysis

In [1]:
import json, time, requests, urllib.parse
import pandas as pd
from tqdm.autonotebook import tqdm

import logging
logging.basicConfig(filename="errors.log", level=logging.ERROR)

/var/folders/rd/v2v17qs50rz2m3cx4yv9ftl40000gn/T/ipykernel_96536/1554299361.py:3: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from tqdm.autonotebook import tqdm


## Definining Article Page Info MediaWiki API parameter and function 

In [2]:
#########
#
#    CONSTANTS
#

# The basic English Wikipedia API endpoint
API_ENWIKIPEDIA_ENDPOINT = "https://en.wikipedia.org/w/api.php"

# We'll assume that there needs to be some throttling for these requests - we should always be nice to a free data resource
API_LATENCY_ASSUMED = 0.002       # Assuming roughly 2ms latency on the API and network
API_THROTTLE_WAIT = (1.0/100.0)-API_LATENCY_ASSUMED

# When making automated requests we should include something that is unique to the person making the request
# This should include an email - your UW email would be good to put in there
REQUEST_HEADERS = {
    'User-Agent': '<hbaghar@uw.edu>, University of Washington, MSDS DATA 512 - AUTUMN 2022',
}

# This is just a list of English Wikipedia article titles that we can use for example requests
ARTICLE_TITLES = [ 'Bison', 'Northern flicker', 'Red squirrel', 'Chinook salmon', 'Horseshoe bat' ]

# This is a string of additional page properties that can be returned see the Info documentation for
# what can be included. If you don't want any this can simply be the empty string
PAGEINFO_EXTENDED_PROPERTIES = "talkid|url|watched|watchers"
#PAGEINFO_EXTENDED_PROPERTIES = ""

# This template lists the basic parameters for making this
PAGEINFO_PARAMS_TEMPLATE = {
    "action": "query",
    "format": "json",
    "titles": "",           # to simplify this should be a single page title at a time
    "prop": "info",
    "inprop": PAGEINFO_EXTENDED_PROPERTIES
}


In [3]:
#########
#
#    PROCEDURES/FUNCTIONS
#

def request_pageinfo_per_article(article_title = None, 
                                 endpoint_url = API_ENWIKIPEDIA_ENDPOINT, 
                                 request_template = PAGEINFO_PARAMS_TEMPLATE,
                                 headers = REQUEST_HEADERS):
    # Make sure we have an article title
    if not article_title: return None
    
    request_template['titles'] = article_title
        
    # make the request
    try:
        # we'll wait first, to make sure we don't exceed the limit in the situation where an exception
        # occurs during the request processing - throttling is always a good practice with a free
        # data source like Wikipedia - or any other community sources
        if API_THROTTLE_WAIT > 0.0:
            time.sleep(API_THROTTLE_WAIT)
        response = requests.get(endpoint_url, headers=headers, params=request_template)
        json_response = response.json()
    except Exception as e:
        logging.error("Error making request for article title: {}".format(article_title))
        json_response = None
    return json_response

## Defining ORES API parameters and function

In [4]:
#########
#
#    CONSTANTS
#

# The current ORES API endpoint
API_ORES_SCORE_ENDPOINT = "https://ores.wikimedia.org/v3"
# A template for mapping to the URL
API_ORES_SCORE_PARAMS = "/scores/{context}/{revid}/{model}"

# Use some delays so that we do not hammer the API with our requests
API_LATENCY_ASSUMED = 0.002       # Assuming roughly 2ms latency on the API and network
API_THROTTLE_WAIT = (1.0/100.0)-API_LATENCY_ASSUMED

# When making automated requests we should include something that is unique to the person making the request
# This should include an email - your UW email would be good to put in there
REQUEST_HEADERS = {
    'User-Agent': '<hbaghar@uw.edu>, University of Washington, MSDS DATA 512 - AUTUMN 2022'
}

# A dictionary of English Wikipedia article titles (keys) and sample revision IDs that can be used for this ORES scoring example
ARTICLE_REVISIONS = { 'Bison':1085687913 , 'Northern flicker':1086582504 , 'Red squirrel':1083787665 , 'Chinook salmon':1085406228 , 'Horseshoe bat':1060601936 }

# This template lists the basic parameters for making an ORES request
ORES_PARAMS_TEMPLATE = {
    "context": "enwiki",        # which WMF project for the specified revid
    "revid" : "",               # the revision to be scored - this will probably change each call
    "model": "articlequality"   # the AI/ML scoring model to apply to the reviewion
}
#
# The current ML models for English wikipedia are:
#   "articlequality"
#   "articletopic"
#   "damaging"
#   "version"
#   "draftquality"
#   "drafttopic"
#   "goodfaith"
#   "wp10"
#
# The specific documentation on these is scattered so if you want to use one you'll have to look around.
#

In [5]:
#########
#
#    PROCEDURES/FUNCTIONS
#

def request_ores_score_per_article(article_revid = None, 
                                   endpoint_url = API_ORES_SCORE_ENDPOINT, 
                                   endpoint_params = API_ORES_SCORE_PARAMS, 
                                   request_template = ORES_PARAMS_TEMPLATE,
                                   headers = REQUEST_HEADERS,
                                   features=False):
    # Make sure we have an article revision id
    if not article_revid: return None
    
    # set the revision id into the template
    request_template['revid'] = article_revid
    
    # now, create a request URL by combining the endpoint_url with the parameters for the request
    request_url = endpoint_url+endpoint_params.format(**request_template)
    
    # the features used by the ML model can sometimes be returned as well as scores
    if features:
        request_url = request_url+"?features=true"
    
    # make the request
    try:
        # we'll wait first, to make sure we don't exceed the limit in the situation where an exception
        # occurs during the request processing - throttling is always a good practice with a free
        # data source like ORES - or other community sources
        if API_THROTTLE_WAIT > 0.0:
            time.sleep(API_THROTTLE_WAIT)
        response = requests.get(request_url, headers=headers)
        json_response = response.json()
    except Exception as e:
        json_response = None
    return json_response

# 1. Getting Article and Population Data

The first step is getting the data, which lives in several different places. You will need data that lists Wikipedia articles of politicians and data for country populations.
The Wikipedia Category:Politicians by nationality was crawled to generate a list of Wikipedia article pages about politicians from a wide range of countries. This data is in the homework folder as [`politicians_by_country.SEPT.2022.csv`](https://docs.google.com/spreadsheets/u/0/d/1Y4vSTYENgNE5KltqKZqnRQQBQZN5c8uKbSM4QTt8QGg/edit).

The population data is available in CSV format as [`population_by_country_2022.csv`](https://docs.google.com/spreadsheets/u/0/d/1POuZDfA1sRooBq9e1RNukxyzHZZ-nQ2r6H5NcXhsMPU/edit) from the homework folder. This dataset is drawn from the world population data sheet published by the Population Reference Bureau.

**Some Considerations**

You should be a little careful with the data. Crawling Wikipedia categories to identify relevant page subsets can result in misleading and/or duplicate category labels. Naturally, the data crawl attempted to resolve these, but not all may have been caught. You should document how you handle any data inconsistencies.

The population_by_country_2022.csv contains some rows that provide cumulative regional population counts. These rows are distinguished by having ALL CAPS values in the 'geography' field (e.g. AFRICA, OCEANIA). These rows won't match the country values in `politicians_by_country.SEPT.2022.csv`, but you will want to retain some of them so that you can report coverage and quality by region as specified in the analysis section below.


## 1.1 Loading list of articles that need to be scored

In [6]:
articles = pd.read_csv("https://docs.google.com/spreadsheets/d/1Y4vSTYENgNE5KltqKZqnRQQBQZN5c8uKbSM4QTt8QGg/export?gid=1672307727&format=csv")
articles.head()

name                                                url  \
0        Shahjahan Noori      https://en.wikipedia.org/wiki/Shahjahan_Noori   
1  Abdul Ghafar Lakanwal  https://en.wikipedia.org/wiki/Abdul_Ghafar_Lak...   
2         Majah Ha Adrif       https://en.wikipedia.org/wiki/Majah_Ha_Adrif   
3      Haroon al-Afghani    https://en.wikipedia.org/wiki/Haroon_al-Afghani   
4            Tayyab Agha          https://en.wikipedia.org/wiki/Tayyab_Agha   

       country  
0  Afghanistan  
1  Afghanistan  
2  Afghanistan  
3  Afghanistan  
4  Afghanistan

## 1.2 Loading population dataset and processing it

We want to map regions and countires together instead of the hierarchy being broken down by row in the same column

In [7]:
population = pd.read_csv("https://docs.google.com/spreadsheets/d/1POuZDfA1sRooBq9e1RNukxyzHZZ-nQ2r6H5NcXhsMPU/export?gid=1154770218&format=csv")
population.head()

Geography  Population (millions)
0            WORLD                 7963.0
1           AFRICA                 1419.0
2  NORTHERN AFRICA                  251.0
3          Algeria                   44.9
4            Egypt                  103.5

In [8]:
population["Region"] = population["Geography"]
population.loc[~(population["Region"].str.isupper()), "Region"] = pd.NA
population["Region"] = population["Region"].fillna(method="ffill")
population.rename({"Geography": "Country"}, axis=1, inplace=True)
population_totals = population.loc[(population["Region"] == population["Country"]), ["Region", "Population (millions)"]]
population = population.loc[~((population["Region"] == population["Country"]))]
population = population[["Region","Country","Population (millions)"]]
population.head()

Region  Country  Population (millions)
3  NORTHERN AFRICA  Algeria                   44.9
4  NORTHERN AFRICA    Egypt                  103.5
5  NORTHERN AFRICA    Libya                    6.8
6  NORTHERN AFRICA  Morocco                   36.7
7  NORTHERN AFRICA    Sudan                   46.9

In [9]:
population_totals.head()

Region  Population (millions)
0             WORLD                 7963.0
1            AFRICA                 1419.0
2   NORTHERN AFRICA                  251.0
10   WESTERN AFRICA                  430.0
27   EASTERN AFRICA                  473.0

## 1.3 Getting the Most Recent Revision ID From The Article Page API

In [10]:
for name in tqdm(articles["name"]):
    try:
        page = request_pageinfo_per_article(name)
        pageid = str(list(page["query"]["pages"].keys())[0])
        revid = str(page["query"]["pages"][pageid]["lastrevid"])
        articles.loc[articles["name"] == name, "pageid"] = pageid
        articles.loc[articles["name"] == name, "revid"] = revid
    except KeyError:
        logging.error("Revision ID not found for article: {}".format(name))
    except Exception as e:
        logging.error("Error making request for article: {}. Error raised: {}".format(name, e.__cause__))

100%|██████████| 7584/7584 [28:59<00:00,  4.36it/s]


# 2. Get Article Quality Predictions

Now you need to get the predicted quality scores for each article in the Wikipedia dataset. We're using a machine learning system called ORES. This was originally an acronym for "Objective Revision Evaluation Service" but was simply renamed “ORES”. ORES is a machine learning tool that can provide estimates of Wikipedia article quality. The article quality estimates are, from best to worst:

1. FA - Featured article
2. GA - Good article
3. B - B-class article
4. C - C-class article
5. Start - Start-class article
6. Stub - Stub-class article


In [11]:
for name,revid in tqdm(zip(articles["name"],articles["revid"])):
    try:
        ores = request_ores_score_per_article(revid)
        ores = ores["enwiki"]["scores"][revid]["articlequality"]["score"]["prediction"]
        articles.loc[articles["revid"] == revid, "ores"] = ores
    except Exception as e:
        logging.error("ORES score not found for article: {}. Error raised: {}".format(name, e.__cause__))

articles.to_csv("articles.csv", index=False)

7584it [25:56,  4.87it/s]


In [13]:
articles.isna().mean()*100

name       0.0000
url        0.0000
country    0.0000
pageid     0.0923
revid      0.0923
ores       0.0923
dtype: float64

We see 0.0923% missing data points, we can choose not to resolve these errors as it is a very tiny set of pages.

# 3. Combining Datasets

Some processing of the data will be necessary! In particular, you'll need to - after retrieving and including the ORES data for each article - merge the wikipedia data and population data together. Both have fields containing country names for just that purpose. 

After merging the data, you'll invariably run into entries which cannot be merged. Either the population dataset does not have an entry for the equivalent Wikipedia country, or vice-versa.


Identify all countries for which there are no matches and output a list of those countries, with each country on a separate line called:
`wp_countries-no_match.txt`

Consolidate the remaining data into a single CSV file called:
`wp_politicians_by_country.csv`

The schema for that file should look something like this:

|Column|
|---|
|country|
|region|
|population|
|article_title|
|revision_id|
|article_quality|



In [131]:
articles = pd.read_csv("articles.csv")

In [132]:
combined_data = pd.merge(articles, population, how="outer", left_on="country", right_on="Country")
combined_data.shape

(7609, 9)

In [133]:
combined_data.isna().sum()

name                     25
url                      25
country                  25
pageid                   32
revid                    32
ores                     32
Region                   70
Country                  70
Population (millions)    70
dtype: int64

In [134]:
# In left table only: Checking for countries with missing left table fields and valid Country (right table) field
left_unmatched = combined_data.loc[combined_data["country"].isna(), "Country"].unique()

# In right table only: Checking for countries with missing right table fields and valid country (left table) field
right_unmatched = combined_data.loc[combined_data["Country"].isna(), "country"].unique()

In [135]:
import numpy as np

with open('wp_countries-no_match.txt', 'w') as f:
    for country in np.hstack((left_unmatched, right_unmatched)):
        f.write(country+"\n")


In [136]:
combined_data = combined_data.dropna(axis=0)
combined_data.rename(
    {
        "Population (millions)": "population",
        "Region":"region", 
        "revid":"revision_id", 
        "ores":"article_quality",
        "name":"article_title"
    }, axis=1, inplace=True)
combined_data[["country", "region", "population", "article_title", "revision_id", "article_quality"]].to_csv("wp_politicians_by_country.csv", index=False)

# 4. Analysis

Your analysis will consist of calculating total-articles-per-population (a ratio representing the number of articles per person)  and high-quality-articles-per-population (a ratio representing the number of high quality articles per person) on a country-by-country and regional basis. All of these values are to be “per capita”.

In this analysis a country can only exist in one region. The `population_by_country_2022.csv` actually represents regions in a hierarchical order. For your analysis always put a country in the closest (lowest in the hierarchy) region.
For this analysis you should consider "high quality" articles to be articles that ORES predicted would be in either the "FA" (featured article) or "GA" (good article) classes.

Also, keep in mind that the `population_by_country_2022.csv` file provides population in millions. The calculated proportions in this step are likely to be very small numbers.


## 4.1 Total Articles per Capita

In [137]:
combined_data = combined_data[["country", "region", "population", "article_title", "revision_id", "article_quality"]]
total_articles_by_country = combined_data.groupby("country").agg({"article_title":"count", "population":"max"}).reset_index()
total_articles_by_country.set_index("country", inplace=True)
total_articles_by_country

article_title  population
country                               
Afghanistan            118        41.1
Albania                 83         2.8
Algeria                 34        44.9
Andorra                 10         0.1
Angola                  42        35.6
...                    ...         ...
Venezuela               62        28.3
Vietnam                 27        99.4
Yemen                   61        33.7
Zambia                  13        20.0
Zimbabwe                63        16.3

[184 rows x 2 columns]

In [138]:
total_articles_by_region = combined_data.groupby("region").agg({"article_title":"count"}).merge(population_totals, how="left", left_index=True, right_on="Region")
total_articles_by_region.set_index("Region", inplace=True)
total_articles_by_region

article_title  Population (millions)
Region                                               
CARIBBEAN                  201                   44.0
CENTRAL AMERICA            195                  178.0
CENTRAL ASIA               106                   78.0
EAST ASIA                  245                 1674.0
EASTERN AFRICA             650                  473.0
EASTERN EUROPE             735                  287.0
MIDDLE AFRICA              203                  196.0
NORTHERN AFRICA            227                  251.0
NORTHERN EUROPE            262                  107.0
OCEANIA                     86                   44.0
SOUTH AMERICA              577                  434.0
SOUTH ASIA                 648                 2008.0
SOUTHEAST ASIA             412                  676.0
SOUTHERN AFRICA            117                   69.0
SOUTHERN EUROPE            890                  151.0
WESTERN AFRICA             568                  430.0
WESTERN ASIA               686                  294.0
WESTERN EUROPE             699                  197.0

## 4.2 High Quality Articles per Capita

In [139]:
high_qual_counts = lambda x: x[x.isin(["FA","GA"])].shape[0]
high_quality_articles_by_country = combined_data.groupby("country").agg({"article_quality":high_qual_counts, "population":"max"})
high_quality_articles_by_country

article_quality  population
country                                 
Afghanistan                6        41.1
Albania                    6         2.8
Algeria                    0        44.9
Andorra                    2         0.1
Angola                     0        35.6
...                      ...         ...
Venezuela                  0        28.3
Vietnam                    2        99.4
Yemen                      2        33.7
Zambia                     0        20.0
Zimbabwe                   0        16.3

[184 rows x 2 columns]

In [140]:
high_quality_articles_by_region = combined_data.groupby("region").agg({"article_quality":high_qual_counts}).merge(population_totals, how="left", left_index=True, right_on="Region")
high_quality_articles_by_region.set_index("Region", inplace=True)
high_quality_articles_by_region

article_quality  Population (millions)
Region                                                 
CARIBBEAN                      8                   44.0
CENTRAL AMERICA               10                  178.0
CENTRAL ASIA                   3                   78.0
EAST ASIA                     16                 1674.0
EASTERN AFRICA                15                  473.0
EASTERN EUROPE                39                  287.0
MIDDLE AFRICA                  5                  196.0
NORTHERN AFRICA                6                  251.0
NORTHERN EUROPE                8                  107.0
OCEANIA                        2                   44.0
SOUTH AMERICA                 13                  434.0
SOUTH ASIA                    23                 2008.0
SOUTHEAST ASIA                24                  676.0
SOUTHERN AFRICA                4                   69.0
SOUTHERN EUROPE               46                  151.0
WESTERN AFRICA                13                  430.0
WESTERN ASIA                  28                  294.0
WESTERN EUROPE                22                  197.0

# 5. Results



## 5.1 Top 10 countries by coverage

The 10 countries with the highest total articles per capita (in descending order)

In [141]:
total_articles_by_country["article_per_capita"] = total_articles_by_country["article_title"]/(total_articles_by_country["population"]*1e6)
total_articles_by_country = total_articles_by_country.loc[total_articles_by_country["article_per_capita"] != np.inf,:]
total_articles_by_country["article_per_capita"].sort_values(ascending=False).head(10)

country
Antigua and Barbuda               0.000170
Federated States of Micronesia    0.000130
Andorra                           0.000100
Barbados                          0.000093
Marshall Islands                  0.000090
Montenegro                        0.000060
Seychelles                        0.000060
Luxembourg                        0.000053
Bhutan                            0.000051
Grenada                           0.000050
Name: article_per_capita, dtype: float64

## 5.2 Bottom 10 countries by coverage

The 10 countries with the lowest total articles per capita (in ascending order) 

In [142]:
total_articles_by_country["article_per_capita"].sort_values(ascending=True).head(10)

country
China           1.392176e-09
Mexico          7.843137e-09
Saudi Arabia    8.174387e-08
Romania         1.052632e-07
India           1.255998e-07
Sri Lanka       1.339286e-07
Egypt           1.352657e-07
Ethiopia        2.025932e-07
Taiwan          2.155172e-07
Vietnam         2.716298e-07
Name: article_per_capita, dtype: float64

## 5.3 Top 10 countries by high quality

The 10 countries with the highest high quality articles per capita (in descending order)

In [143]:
high_quality_articles_by_country["article_per_capita"] = high_quality_articles_by_country["article_quality"]/(high_quality_articles_by_country["population"]*1e6)
high_quality_articles_by_country = high_quality_articles_by_country.loc[high_quality_articles_by_country["article_per_capita"] != np.inf,:]
high_quality_articles_by_country["article_per_capita"].sort_values(ascending=False).head(10)

country
Andorra                  2.000000e-05
Montenegro               5.000000e-06
Albania                  2.142857e-06
Suriname                 1.666667e-06
Bosnia-Herzegovina       1.470588e-06
Lithuania                1.071429e-06
Croatia                  1.052632e-06
Slovenia                 9.523810e-07
Palestinian Territory    9.259259e-07
Gabon                    8.333333e-07
Name: article_per_capita, dtype: float64

## 5.4 Bottom 10 countries by high quality

The 10 countries with the lowest high quality articles per capita (in ascending order)

In [144]:
high_quality_articles_by_country["article_per_capita"].sort_values(ascending=True).head(10)

country
Zimbabwe                          0.0
Zambia                            0.0
Kuwait                            0.0
Sri Lanka                         0.0
St. Kitts-Nevis                   0.0
Jamaica                           0.0
Italy                             0.0
Israel                            0.0
Iceland                           0.0
St. Vincent and the Grenadines    0.0
Name: article_per_capita, dtype: float64

We should check if there are more than 10 countries with zeros.

In [155]:
high_quality_articles_by_country[high_quality_articles_by_country["article_per_capita"] == 0].index.unique()

Index(['Algeria', 'Angola', 'Antigua and Barbuda', 'Argentina', 'Austria',
       'Bahamas', 'Bahrain', 'Bangladesh', 'Barbados', 'Belize', 'Bhutan',
       'Botswana', 'Brazil', 'Cape Verde', 'Chad', 'China', 'Comoros', 'Congo',
       'Congo, Dem. Rep.', 'Cote d'Ivoire', 'Cyprus', 'Denmark', 'Djibouti',
       'Dominica', 'Egypt', 'Equatorial Guinea', 'Eritrea',
       'Federated States of Micronesia', 'Fiji', 'Finland', 'Gambia', 'Ghana',
       'Greece', 'Grenada', 'Guinea-Bissau', 'Guyana', 'Honduras', 'Iceland',
       'Israel', 'Italy', 'Jamaica', 'Kuwait', 'Laos', 'Latvia', 'Lesotho',
       'Liberia', 'Luxembourg', 'Madagascar', 'Malawi', 'Maldives', 'Malta',
       'Marshall Islands', 'Mexico', 'Moldova', 'Mongolia', 'Mozambique',
       'Namibia', 'Nicaragua', 'Niger', 'North Macedonia', 'Oman', 'Paraguay',
       'Qatar', 'Saint Lucia', 'Samoa', 'Seychelles', 'Sierra Leone',
       'Singapore', 'Solomon Islands', 'Sri Lanka', 'St. Kitts-Nevis',
       'St. Vincent and the G

We should also check lowest scores which are not zero.

In [153]:
high_quality_articles_by_country[high_quality_articles_by_country["article_per_capita"] > 0]["article_per_capita"].sort_values(ascending=True).head(10)

country
India       4.233700e-09
Thailand    1.497006e-08
Japan       1.601281e-08
Nigeria     1.830664e-08
Vietnam     2.012072e-08
Colombia    2.036660e-08
Uganda      2.118644e-08
Pakistan    2.120441e-08
Sudan       2.132196e-08
Iran        2.257336e-08
Name: article_per_capita, dtype: float64

## 5.5 Geographic regions by total coverage

A rank ordered list of geographic regions (in descending order) by total articles per capita

In [145]:
total_articles_by_region["article_per_capita"] = total_articles_by_region["article_title"]/(total_articles_by_region["Population (millions)"]*1e6)
total_articles_by_region = total_articles_by_region.loc[total_articles_by_region["article_per_capita"] != np.inf,:]
total_articles_by_region["article_per_capita"].sort_values(ascending=False).head(10)

Region
SOUTHERN EUROPE    0.000006
CARIBBEAN          0.000005
WESTERN EUROPE     0.000004
EASTERN EUROPE     0.000003
NORTHERN EUROPE    0.000002
WESTERN ASIA       0.000002
OCEANIA            0.000002
SOUTHERN AFRICA    0.000002
EASTERN AFRICA     0.000001
CENTRAL ASIA       0.000001
Name: article_per_capita, dtype: float64

## 5.6 Geographic regions by high quality coverage

Rank ordered list of geographic regions (in descending order) by high quality articles per capita

In [146]:
high_quality_articles_by_region["article_per_capita"] = high_quality_articles_by_region["article_quality"]/(high_quality_articles_by_region["Population (millions)"]*1e6)
high_quality_articles_by_region = high_quality_articles_by_region.loc[high_quality_articles_by_region["article_per_capita"] != np.inf,:]
high_quality_articles_by_region["article_per_capita"].sort_values(ascending=False).head(10)

Region
SOUTHERN EUROPE    3.046358e-07
CARIBBEAN          1.818182e-07
EASTERN EUROPE     1.358885e-07
WESTERN EUROPE     1.116751e-07
WESTERN ASIA       9.523810e-08
NORTHERN EUROPE    7.476636e-08
SOUTHERN AFRICA    5.797101e-08
CENTRAL AMERICA    5.617978e-08
OCEANIA            4.545455e-08
CENTRAL ASIA       3.846154e-08
Name: article_per_capita, dtype: float64